### PR TITLE
WIP [DeepSeek R1] Add DeepSeekV3 Base + Weight Conversion

### PR DESCRIPTION
**Pull request by DavidLandup0**
Adds DeepSeekV3 base and weight conversion script.

The architecture itself builds and runs, but requires massive RAM. Example of a one-block model running on some tokens below (5s/token):

image

Needs more refactoring and simplification.

### WIP/TODOs
- The weight download takes around 880GB of disk space. Then instantiating the model and torch weights both in memory requires massive RAM. Figure out if this can be done iteratively.
- Figure out Keras weight sharding + this
- Move from ModelArgs dataclass syntax into config.json style config

### CODE REVIEW

File: keras_hub/src/models/deepseek_r1/deepseek_backbone.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            mask = ops.full((seqlen, seqlen), float("-inf"))
+            mask = ops.triu(mask, k=1)
+        for layer in self.blocks:
+            h = layer(h, start_pos=start_pos, freqs_cis=freqs_cis, mask=mask)
+        h = self.norm(h)[:, -1]
+        logits = self.head(h)
+        return logits
+
+
+if __name__ == "__main__":
```
Comment (DavidLandup0) [inline keras_hub/src/models/deepseek_r1/deepseek_backbone.py:208]: Sanity check main call - will be removed.

File: keras_hub/src/models/deepseek_r1/deepseek_layers.py [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+from keras import layers
+from keras import ops
+
+from keras_hub.src.layers.modeling.rms_normalization import RMSNormalization
+
+attn_impl = "absorb"
+rank = 0
+
+
+class Embedding(layers.Layer):
```
Comment (DavidLandup0) [inline keras_hub/src/models/deepseek_r1/deepseek_layers.py:13]: TODO: Remove custom class and just use layers.Embedding.

File: keras_hub/src/models/deepseek_r1/deepseek_layers.py [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+            torch.empty(
+                scale_out_features, scale_in_features, dtype=torch.float32
+            )
+        )
+
+    def call(self, x):
+        return linear(x, self.weight, self.bias)
+
+
+class ColumnParallelLinear(Linear):
```
Comment (DavidLandup0) [inline keras_hub/src/models/deepseek_r1/deepseek_layers.py:95]: Do we still need the custom `XParallel` classes if we don't use torch.dist which boils most of them back to the standard implementations?

File: keras_hub/api/__init__.py [hunk 4]
```diff
@@ -4,12 +4,12 @@
...
 """
 
-from keras_hub.api import layers
-from keras_hub.api import metrics
-from keras_hub.api import models
-from keras_hub.api import samplers
-from keras_hub.api import tokenizers
-from keras_hub.api import utils
-from keras_hub.src.utils.preset_utils import upload_preset
+from keras_hub.api import layers as layers
```
Comment (sachinprasadhs) [inline keras_hub/api/__init__.py:7]: Can you rebase the code to latest changes in Keras Hub in master and generate API again, so that it will show only the changes from DeepSeek

File: keras_hub/src/models/deepseek_r1/deepseek/config.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
+{
+    "module": "keras_hub.models.DeepSeekV3Backbone",
+    "class_name": "DeepSeekV3Backbone",
+    "config": {
+      "max_batch_size": 1,
+      "max_seq_len": 16384,
+      "vocab_size": 129280,
+      "dim": 7168,
+      "inter_dim": 10944,
+      "moe_inter_dim": 1408,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_r1/deepseek/config.json:10]: You can remove both the config files and the directory deepseek, this will be part of the checkpoint files which will be uploaded.

File: keras_hub/src/models/deepseek_r1/deepseek_backbone.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        # yarn
+        original_seq_len,
+        rope_theta,
+        rope_factor,
+        beta_fast,
+        beta_slow,
+        mscale,
+        **kwargs,
+    ):
+        # === Layers ===
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_r1/deepseek_backbone.py:99]: Replace all the n_<names> to num_<names>, example n_layers --> num_layers, so that this will be consistent with other implementaions.

File: keras_hub/src/models/deepseek_r1/deepseek_causallm_preprocessor.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        sequence_length: Pass to override the configured `sequence_length` of
+            the layer.
+
+    Examples:
+    ```python
+    # Load the preprocessor from a preset.
+    preprocessor = keras_hub.models.DeepSeekR1CausalLMPreprocessor.from_preset(
+        "..."
+    )
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_r1/deepseek_causallm_preprocessor.py:49]: You can mention the probable preset name, the same thing can be used in conversion script/ Kaggle whereever it is applicable.

File: keras_hub/src/models/deepseek_r1/deepseek_layers.py [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+    # flatten on axis 3
+    y = ops.reshape(y, x.shape[:3] + (-1,))
+    return y
+
+
+class MLA(layers.Layer):
+    """
+    Multi-Headed Attention Layer (MLA).
+
+    Attributes:
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_r1/deepseek_layers.py:282]: Attributes --> Args

File: keras_hub/src/models/deepseek_r1/deepseek_layers.py [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+    freqs_cis = ops.reshape(freqs_cis, [1, x.shape[1], 1, x.shape[-1]])
+    y = _view_as_real(x * freqs_cis)
+    # flatten on axis 3
+    y = ops.reshape(y, x.shape[:3] + (-1,))
+    return y
+
+
+class MLA(layers.Layer):
+    """
+    Multi-Headed Attention Layer (MLA).
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_r1/deepseek_layers.py:280]: change this to --> """Multi-Headed Attention Layer (MLA).

 Basically, """ should be followed by string in the same line

**Review by sachinprasadhs — COMMENTED** [review]
Did one normal pass of review and have made generic comments.

**Comment by divyashreepathihalli** [comment]
@DavidLandup0 - any updates here?
